File: 02_arrays/isSorted/isSorted.cpp
#include <iostream>
using namespace std;

// Recursive way - 
bool sorted(vector<int>& arr, int n){
    if(n<=1)return true;
    if(arr[0] > arr[1]) return false;
    return sorted(arr +1, n-1);
} 
bool isSorted(vector<int>& arr) {
    // code here
    int n = arr.size();
    return sorted(arr, n);
}

// Iterative way - 
bool isSorted(vector<int>& arr) {
    
    // Iterate over the array and check if 
    // every element is greater than or
    // equal to previous element.
    for (int i = 1; i < arr.size(); i++)
        if (arr[i - 1] > arr[i])
            return false;

    return true;
}

// Using Inbuilt function 
bool isSorted(vector<int> &arr){
    
    if(is_sorted(arr.begin(), arr.end()))
    return true;
    
    return false;
}

File: 02_arrays/isSorted/isSorted.java
import java.util.*;

public class Main {

    // ================= RECURSIVE WAY =================
    static boolean sorted(int[] arr, int index, int n) {
        if (n <= 1) return true;

        if (arr[index] > arr[index + 1])
            return false;

        return sorted(arr, index + 1, n - 1);
    }

    static boolean isSortedRecursive(int[] arr) {
        return sorted(arr, 0, arr.length);
    }


    // ================= ITERATIVE WAY =================
    static boolean isSortedIterative(int[] arr) {
        for (int i = 1; i < arr.length; i++) {
            if (arr[i - 1] > arr[i])
                return false;
        }
        return true;
    }


    // ================= USING INBUILT FUNCTION =================
    static boolean isSortedInbuilt(int[] arr) {
        // Java has no direct is_sorted, so compare with sorted copy
        int[] copy = arr.clone();
        Arrays.sort(copy);
        return Arrays.equals(arr, copy);
    }


    // ================= MAIN =================
    public static void main(String[] args) {

        int[] arr = {1, 2, 3, 4, 5};

        System.out.println("Recursive: " + isSortedRecursive(arr));
        System.out.println("Iterative: " + isSortedIterative(arr));
        System.out.println("Inbuilt: " + isSortedInbuilt(arr));
    }
}
